pallets/avn-transaction-payment/src/tests/adjustment_fee_tests.rs

```diff
@@ -0,0 +1,582 @@
+use super::*;
+use crate::mock::{
+    event_emitted, new_test_ext, roll_one_block, AccountId, Balances, RuntimeCall, RuntimeEvent,
+    RuntimeOrigin, System, TestAccount, TestRuntime,
+};
+
+use frame_support::{dispatch::DispatchInfo, pallet_prelude::Weight};
+use pallet_transaction_payment::ChargeTransactionPayment;
+use sp_runtime::traits::SignedExtension;
+
+use frame_support::assert_ok;
+
+pub const TX_LEN: usize = 1;
+pub const INITIAL_SENDER_BALANCE: u128 = 200;
+pub const BASE_FEE: u32 = 14;
+pub const FIXED_FEE: u128 = 10;
+pub const PERCENTAGE_FEE: u32 = 50;
+
+fn to_acc_id(id: u64) -> AccountId {
+    return TestAccount::new(id).account_id()
+}
+
+/// create a transaction info struct from weight. Handy to avoid building the whole struct.
+pub fn info_from_weight(w: Weight) -> DispatchInfo {
+    DispatchInfo { weight: w, ..Default::default() }
+}
+
+fn pay_gas_and_call_remark(sender: &AccountId) {
+    let pre = <ChargeTransactionPayment<TestRuntime> as SignedExtension>::pre_dispatch(
+        ChargeTransactionPayment::from(0),
+        sender,
+        &RuntimeCall::System(frame_system::Call::remark { remark: vec![] }),
+        &info_from_weight(Weight::from_ref_time(1)),
+        TX_LEN,
+    );
+
+    assert_ok!(&pre);
+
+    assert_ok!(System::remark(RuntimeOrigin::signed(*sender), vec![])
+        .map_err(|_e| Error::<TestRuntime>::InvalidFeeConfig));
+
+    assert_ok!(ChargeTransactionPayment::<TestRuntime>::post_dispatch(
+        Some(pre.expect("Checked for error")),
+        &DispatchInfo { weight: Weight::from_ref_time(1), ..Default::default() },
+        &PostDispatchInfo { actual_weight: None, pays_fee: Default::default() },
+        TX_LEN,
+        &Ok(())
+    ));
+
+    System::inc_account_nonce(sender);
+}
+
+fn set_initial_sender_balance(sender: &AccountId) {
+    Balances::make_free_balance_be(&sender, INITIAL_SENDER_BALANCE);
+    assert_eq!(INITIAL_SENDER_BALANCE, Balances::free_balance(sender));
+}
+
+fn get_percentage_fee_value(percentage_fee: u32) -> u32 {
+    (BASE_FEE * percentage_fee) / 100
```

Comment:
Shouldn't there be a test here for when this rate is invalid, that is, outside the [0;1] range?
It should not be negative or larger than 1, otherwise we'll get very odd behaviour.

Reply:
There are tests for that in the set_know_sender_tests suite!